packages/planar-shadow/src/PlanarShadow.ts

```diff
@@ -0,0 +1,190 @@
+import {
+  BaseMaterial,
+  BlendFactor,
+  Color,
+  CompareFunction,
+  Engine,
+  Shader,
+  StencilOperation,
+  Vector3
+} from "oasis-engine";
+
+/**
+ * Planar Shadow
+ */
+export class PlanarShadow extends BaseMaterial {
+  private static _lightDirProp = Shader.getPropertyByName("u_lightDir");
+  private static _planarHeightProp = Shader.getPropertyByName("u_planarHeight");
+  private static _shadowColorProp = Shader.getPropertyByName("u_planarShadowColor");
+  private static _shadowFalloffProp = Shader.getPropertyByName("u_planarShadowFalloff");
+
+  /**
+   * Planar height
+   */
+  get planarHeight(): number {
+    return this.shaderData.getFloat(PlanarShadow._planarHeightProp);
+  }
+
+  set planarHeight(value: number) {
+    this.shaderData.setFloat(PlanarShadow._planarHeightProp, value);
+  }
+
+  /**
+   * Light direction
+   */
+  get LightDir(): Vector3 {
+    return this.shaderData.getVector3(PlanarShadow._lightDirProp);
+  }
+
+  set LightDir(value: Vector3) {
+    const lightDir = this.shaderData.getVector3(PlanarShadow._lightDirProp);
+    if (value !== lightDir) {
+      value.normalize().cloneTo(lightDir);
+    } else {
+      value.normalize();
+    }
+  }
+
+  /**
+   * Shadow color
+   */
+  get shadowColor(): Color {
+    return this.shaderData.getColor(PlanarShadow._shadowColorProp);
+  }
+
+  set shadowColor(value: Color) {
+    const shadowColor = this.shaderData.getColor(PlanarShadow._shadowColorProp);
+    if (value !== shadowColor) {
+      value.cloneTo(shadowColor);
+    }
+  }
+
+  /**
+   * Shadow falloff coefficient
+   */
+  get shadowFalloff(): number {
+    return this.shaderData.getFloat(PlanarShadow._shadowFalloffProp);
+  }
+
+  set shadowFalloff(value: number) {
+    this.shaderData.setFloat(PlanarShadow._shadowFalloffProp, value);
+  }
+
+  constructor(engine: Engine) {
+    super(engine, Shader.find("planar-shadow"));
+
+    this.isTransparent = true;
+    this.renderState.stencilState.enabled = true;
+    this.renderState.stencilState.referenceValue = 0;
+    this.renderState.stencilState.compareFunctionFront = CompareFunction.Equal;
+    this.renderState.stencilState.compareFunctionBack = CompareFunction.Equal;
+    this.renderState.stencilState.failOperationFront = StencilOperation.Keep;
+    this.renderState.stencilState.failOperationBack = StencilOperation.Keep;
+    this.renderState.stencilState.zFailOperationFront = StencilOperation.Keep;
+    this.renderState.stencilState.zFailOperationBack = StencilOperation.Keep;
+    this.renderState.stencilState.passOperationFront = StencilOperation.IncrementWrap;
+    this.renderState.stencilState.passOperationBack = StencilOperation.IncrementWrap;
+    this.renderState.blendState.targetBlendState.sourceAlphaBlendFactor = BlendFactor.OneMinusSourceAlpha;
+    this.renderState.blendState.targetBlendState.destinationAlphaBlendFactor = BlendFactor.One;
```

Comment:
const {stencilState, blendState:{targetBlendState}} = this.renderState;
stencilState.***=**;
targetBlendState.**=**;

Reply:
done